### PR TITLE
[cpp_on_shore_task_6]iterator learning

### PR DESCRIPTION
**Pull request by ChristopherCao**
### What?
-brief: 学习了stl iterator的接口、源码以及设计思路, 并编写了测试程序对学习到的内容进行验证
-author: 936481843@qq.com

### Directory Specification
```
├─cpp_on_shore
│  ├─chris
│  │  ├─doc
│  │  │  ├─task_1
│  │  │  ├─task_2
│  │  │  ├─task_3
│  │  │  ├─task_4
│  │  │  ├─task_5
│  │  │  └─task_6
│  │  └─src
│  │      ├─task_1
│  │      ├─task_2
│  │      ├─task_3
│  │      ├─task_4
│  │      ├─task_5
│  │      └─task_6
│  ├─roger_lv
│  │  ├─doc
│  │  │  ├─task_1
│  │  │  ├─task_2
│  │  │  ├─task_3
│  │  │  ├─task_4
│  │  │  ├─task_5
│  │  │  └─task_6
│  │  └─src
│  │      ├─task_1
│  │      ├─task_2
│  │      ├─task_3
│  │      ├─task_4
│  │      ├─task_5
│  │      └─task_6
│  └─sheldon
│      ├─doc
│      │  ├─task_1
│      │  │  └─asset
│      │  ├─task_2
│      │  ├─task_3
│      │  ├─task_4
│      │  ├─task_5
│      │  └─task_6
│      └─src
│          ├─task_1
│          ├─task_2
│          ├─task_3
│          ├─task_4
│          ├─task_5
│          └─task_6
```

### What changes are included in this PR?
在最新同步了whale-starry仓库main分支的基础上, 增加创建task_6学习目录, 对stl iterator源码和设计思路进行学习, 保存相应的学习笔记markdown和代码

### Are these changes tested?
所有的代码均已通过本地测试

### CODE REVIEW

File: stl/cpp_on_shore/roger_lv/doc/task_6/learning.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      /// Distance between iterators is represented as this type.
+      typedef _Distance  difference_type;
+      /// This type represents a pointer-to-value_type.
+      typedef _Pointer   pointer;
+      /// This type represents a reference-to-value_type.
+      typedef _Reference reference;
+    };
+    ```
+
+- 实现起来也很简单，参照不同容器下自己定义的iterator即可，std::vector\<int\>::iterator
```
Comment (Light-City) [inline stl/cpp_on_shore/roger_lv/doc/task_6/learning.md:27]: format?

File: stl/cpp_on_shore/sheldon/src/task_6/main.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
+#include <iostream>
+#include <vector>
+
+template <typename T>
+class MyIterator {
+ public:
+  typedef T value_type;
```
Comment (Light-City) [inline stl/cpp_on_shore/sheldon/src/task_6/main.cpp:7]: maybe use using and typename is better

File: stl/cpp_on_shore/roger_lv/doc/task_6/learning.md (hunk 1 above)
Comment (Rogerlv51) [reply stl/cpp_on_shore/roger_lv/doc/task_6/learning.md]: Corrected

File: stl/cpp_on_shore/sheldon/src/task_6/main.cpp (hunk 2 above)
Comment (SheldonLoveCoding) [reply stl/cpp_on_shore/sheldon/src/task_6/main.cpp:7]: Sorry mentor, I met a bug when I modified the above code to `typedef typename T value_type;`. The bug message is `error: expected nested-name-specifier before 'T'`. 

I googled this bug and found out that this bug occured with gcc, but not MSVC. 

With gcc, this bug showed that after `typename` keyword I should use some nested name of a template parameter, like `typedef typename class::T value_type`. But in my case, it seems not necessary to use `typename`. Did I misunderstand `typename`?

Comment (Light-City) [reply stl/cpp_on_shore/sheldon/src/task_6/main.cpp:7]: yes, in this case , we cann't use typename.

**Pull request merged**